### PR TITLE
Add include_spirv macro

### PR DESCRIPTION
**Pull request by kvark**
Depends on https://github.com/gfx-rs/wgpu/pull/729
Makes out shader loading in the examples a bit tidier.

### CODE REVIEW

File: src/util/mod.rs [hunk 1]
```diff
@@ -1,41 +1,15 @@
...
-            io::ErrorKind::InvalidData,
-            "input missing SPIR-V magic number",
-        ));
-    }
-    Ok(result)
+
+    let (pre, words, post) = unsafe { data.align_to::<u32>() };
+    assert_eq!(pre, &[], "data offset is not aligned to words!");
+    assert_eq!(post, &[], "data size is not aligned to words!");
+    assert_eq!(words[0], MAGIC_NUMBER, "wrong magic word {:x}!", words[0]);
```
Comment (rukai) [inline src/util/mod.rs:13]: Can we extend this to: "wrong magic word {:x}. Make sure you are using a SPIRV file."

File: src/util/mod.rs [hunk 2]
```diff
@@ -1,41 +1,15 @@
-use std::{io, slice};
+#[repr(C)] // guarantee 'bytes' comes after '_align'
+pub struct WorldAligned<Bytes: ?Sized> {
```
Comment (rukai) [inline src/util/mod.rs:2]: This should be WordAligned

File: examples/texture-arrays/main.rs [hunk 3]
```diff
@@ -97,29 +97,26 @@ impl framework::Example for Example {
...
             }
             _ => {
                 panic!("Graphics adapter does not support any of the capabilities needed for this example");
             }
         };
-        let vs_module =
-            device.create_shader_module(&wgpu::read_spirv(std::io::Cursor::new(vs_bytes)).unwrap());
-        let fs_module =
-            device.create_shader_module(&wgpu::read_spirv(std::io::Cursor::new(fs_bytes)).unwrap());
+        let fs_module = device.create_shader_module(wgpu::util::make_spirv(&fs_bytes));
```
Comment (rukai) [inline examples/texture-arrays/main.rs:119]: There is no guarantee that fs_bytes is aligned!

Comment (cwfitzgerald) [reply examples/texture-arrays/main.rs:119]: It's going to be aligned to the alignment of the global memory allocator, which on all platforms we care about is > 4. Wouldn't hurt to add an assert here though.

File: src/lib.rs [hunk 4]
```diff
@@ -571,6 +570,11 @@ impl Drop for ShaderModule {
     }
 }
 
+pub enum ShaderModuleSource<'a> {
```
Comment (cwfitzgerald) [inline src/lib.rs:573]: Could we get some documentation on this?

File: src/util/mod.rs [hunk 5]
```diff
@@ -1,41 +1,15 @@
...
+    pub _align: [u32; 0],
```
Comment (cwfitzgerald) [inline src/util/mod.rs:3]: Is there any reason this struct can't have an explicit alignment attribute?

File: src/util/mod.rs (hunk 1 above)
Comment (cwfitzgerald) [reply src/util/mod.rs:13]: "wrong magic word {:x}. Make sure you are using a **binary** SPIRV file."

File: examples/texture-arrays/main.rs (hunk 3 above)
Comment (rukai) [reply examples/texture-arrays/main.rs:119]: There already is an assert.

So heap memory is automatically aligned to 4.
But stack memory has no such automatic alignment?
I haven't heard of this before.

Comment (cwfitzgerald) [reply examples/texture-arrays/main.rs:119]: Heap memory on linux is aligned to 16, windows 8, don't remember mac but it's probably one of them. Depends on the properties of the allocator, but it's often helpful inside the allocator to have a minimum size that you can allocate.

File: src/util/mod.rs (hunk 2 above)
Comment (kvark) [reply src/util/mod.rs]: ooops

File: src/util/mod.rs (hunk 5 above)
Comment (kvark) [reply src/util/mod.rs]: good idea! I'll try that

**Comment by kvark** [comment]
Thank you both for reviewing! This is updated now, PTAL

File: src/util/mod.rs [hunk 6]
```diff
@@ -1,41 +1,16 @@
-use std::{io, slice};
+#[repr(align(16))]
+pub struct WordAligned<Bytes: ?Sized>(pub Bytes);
```
Comment (cwfitzgerald) [inline src/util/mod.rs:2]: Minor nit: This should either be `align(4)` (which spirv requires) or be called something other than `WordAligned` which implies alignment to 2/4/8s (at least on x86 with a 16/32/64bit word)

File: src/util/mod.rs [hunk 7]
```diff
@@ -1,41 +1,16 @@
...
-    unsafe {
-        // Writing all bytes through a pointer with less strict alignment when our type has no
-        // invalid bitpatterns is safe.
-        x.read_exact(slice::from_raw_parts_mut(
-            result.as_mut_ptr() as *mut u8,
-            words * 4,
-        ))?;
-        result.set_len(words);
-    }
+pub fn make_spirv<'a>(data: &'a [u8]) -> super::ShaderModuleSource<'a> {
```
Comment (rukai) [inline src/util/mod.rs:4]: Considering this is public, we should have documentation specifying data must be 4 aligned.

**Comment by kvark** [comment]
Thank you! Last notes are now addressed.
bors r=cwfitzgerald,rukai

**Pull request merged**